kv.go

```diff
@@ -0,0 +1,387 @@
+		MaxBytes:          cfg.MaxBytes,
+		MaxAge:            cfg.MaxAge,
+		MaxMsgSize:        cfg.MaxValueSize,
+		Storage:           cfg.Storage,
+		Replicas:          replicas,
+	}
+
+	if _, err := jsc.AddStream(scfg); err != nil {
+		return nil, err
+	}
+
+	kv := &kvs{
+		name:   cfg.Bucket,
+		stream: scfg.Name,
+		pre:    fmt.Sprintf(kvSubjectsPreTmpl, cfg.Bucket),
+		nc:     nc,
+		js:     jsc.(*js),
+	}
+	return kv, nil
+}
+
+type kvs struct {
+	name   string
+	stream string
+	pre    string
+	nc     *Conn
+	js     *js
+}
+
+// Get returns the latest value for the key.
+func (kv *kvs) Get(key string) (value []byte, revision uint64, err error) {
+	// Build by hand since simple and avoids stdlib JSON.
+	var b strings.Builder
+	b.WriteString("{\"last_by_subj\":\"")
+	b.WriteString(kv.pre)
+	b.WriteString(key)
+	b.WriteString("\"}")
+
+	o, cancel, err := getJSContextOpts(kv.js.opts)
+	if err != nil {
+		return nil, 0, err
+	}
+	if cancel != nil {
+		defer cancel()
+	}
+
+	// Send request.
+	subj := kv.js.apiSubj(fmt.Sprintf(apiMsgGetT, kv.stream))
+	r, err := kv.nc.RequestWithContext(o.ctx, subj, []byte(b.String()))
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// FIXME(dlc) - Be good to avoid stdlib JSON when possible.
+	// Maybe: https://github.com/goccy/go-json
+	var resp apiMsgGetResponse
+	if err := json.Unmarshal(r.Data, &resp); err != nil {
+		return nil, 0, err
+	}
+	if resp.Error != nil {
+		return nil, 0, errors.New(resp.Error.Description)
+	}
+
+	return resp.Message.Data, resp.Message.Sequence, nil
+}
+
+// Put will place the new value for the key into the store.
+func (kv *kvs) Put(key string, value []byte) (revision uint64, err error) {
+	var b strings.Builder
+	b.WriteString(kv.pre)
+	b.WriteString(key)
+
+	pa, err := kv.js.Publish(b.String(), value)
+	if err != nil {
+		return 0, err
+	}
+	return pa.Sequence, err
+}
+
+// Create will add the key/value pair iff it does not exist.
+func (kv *kvs) Create(key string, value []byte) (revision uint64, err error) {
+	return kv.Update(key, value, 0)
+}
+
+// Update will update the value iff the latest revision matches.
+func (kv *kvs) Update(key string, value []byte, revision uint64) (uint64, error) {
+	var b strings.Builder
+	b.WriteString(kv.pre)
+	b.WriteString(key)
+
+	m := Msg{Subject: b.String(), Header: Header{}, Data: value}
+	m.Header.Set(ExpectedLastSubjSeqHdr, strconv.FormatUint(revision, 10))
+
+	pa, err := kv.js.PublishMsg(&m)
+	if err != nil {
+		return 0, err
+	}
+	return pa.Sequence, err
+}
+
+// Delete the key and all revisions.
+func (kv *kvs) Delete(key string) error {
+	o, cancel, err := getJSContextOpts(kv.js.opts)
+	if err != nil {
+		return err
+	}
+	if cancel != nil {
+		defer cancel()
+	}
+
+	// Build by hand since simple and avoids stdlib JSON.
+	var b strings.Builder
+	b.WriteString("{\"filter\":\"")
+	b.WriteString(kv.pre)
+	b.WriteString(key)
+	b.WriteString("\"}")
+
+	// Send request.
+	subj := kv.js.apiSubj(fmt.Sprintf(apiStreamPurgeT, kv.stream))
```

Comment:
delete works by writing an entry with a delete operation, not by purging

Reply:
https://github.com/nats-io/nats-architecture-and-design/blob/main/adr/ADR-8.md#deleting-values

you put a nil byte message with `KV-Operation: DEL` header.

Reply:
If we want to fully delete the history we can directly purge it's subject

Reply:
Does Purge leave a marker or really make everything go away?

Reply:
Purge is a roll up with delete.

Reply:
How do we delete the delete markers? Do they always just accumulate? I think we place the marker for connected watchers but purge whole stream after.

Reply:
Yeah I don’t think purge will be that common that accumulating tbem will be a problem. Also why I didn’t think purge just deleting everything is a problem - it’s a unusual situation indicating we are done with this key. Shouldn’t have any watchers. 

One use case for purge is if someone just put like 1MB values in filling history and someone just wants to fix that. Then a delete marker is no biggie

Reply:
What I have now is Purge places delete marker through sync js.Publish() then Purges the key in the stream.

Reply:
Roll up didn’t work? We added that feature for this

Reply:
Rollup works like designed. But when you Delete, we place marker, or when you do a Purge we do as well now. But, who deletes the marker?

For now in Purge() I do same as Delete with sync js.PublishMsg() for the empty delete marker to kick connected watchers then I purge complete key including the marker.

Reply:
Delete ads a entry keeping history. Purge is for removing history entirely leaving a tiny marker. 

I had it with no marker. You said that’s completely useless since it breaks the entire concept of a watch. 

So now purge deleted history and leaves a tiny marker. That’s why you added the roll up feature to deliver that in a single shot. 

What you are doing only works if everything is working. It a watch is just disconnected for a short while it gets stuck and unaware leaving us back to square one with broken watchers which you said is entirely unacceptable

Reply:
Having history showing when the purge was done is valuable and I think in line with the general design of hjstoric values as a concept. So I really think the right behaviour is a roll up delete with marker

Reply:
Then we need another call to clear those markers. I think the way I implemented it makes sense, will trigger connected watchers, but purge whole key on Purge including the marker.

On a new Watch if you have local state and you will need to compare anyway for things that have been deleted from your local cache. Most of the time you will have a marker I know.

Even Delete() feels like it should maybe work like Purge. We will get a bunch of folks wondering why they run out of memory doing lots of Puts and Deletes IMO.

Reply:
> Even Delete() feels like it should maybe work like Purge. We will get a bunch of folks wondering why they run out of memory doing lots of Puts and Deletes IMO.

Remember that (history filling up with put/deletes, rathe than just puts) would only happen if the developers specifically set the history depth to something else than the default (1)? You probably wouldn't set history to more than one if you know you're going to be doing puts and deletes (rather than just puts to update a values).

Once you do set history to > 1 then you would understand (would have to be documented too) that deletes are part of the history and therefore that if you do lots of puts and deletes that could fill up your history.
If you set history > 1 then the function of a purge is one of 'delete the value and compact the history' keeping a marker of the purge is then equivalent to 'compacting the history to 1' vs 'compacting the history to 0' (personally given that this purge marker would anyway be very small I'm ok with it being there and compacting history to 1.

By default with history = 1 there functionally no difference between a delete and a purge, history remains at 1 the only difference is you can tell after the fact that a delete vs a purge was used.

Reply:
They are needed to some degree for watchers but also if you want to have an audit trail.

I did ad in PurgeDeletes() if that is needed.